### PR TITLE
Added a comic state machine [#720]

### PR DESCRIPTION
**Pull request by mcpierce**
* Added a backend state machine configuration.
 * Added a column to track the comic's state.

## Status
HOLD

## Does this PR contain migrations?
YES

## Before You Submit Your PR:
- [X] Have you announced your PR on the comixed-dev mailing list?
- [X] New feature (non-breaking change which adds functionality)
- [X] My code follows the code style of this project.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: comixed-services/src/main/java/org/comixedproject/service/comic/ComicService.java [hunk 1]
```diff
@@ -174,32 +174,27 @@ public Comic deleteComic(final long id) throws ComicException {
...
    * @return the updated comic
+   * @throws ComicException if the id is invalid
    */
   @Transactional
-  public Comic updateComic(final long id, final Comic update) {
+  public Comic updateComic(final long id, final Comic update) throws ComicException {
     log.debug("Updating comic: id={}", id);
-
+    final Comic comic = this.doGetComic(id);
     final Optional<Comic> record = this.comicRepository.findById(id);
```
Comment (BRUCELLA2) [inline comixed-services/src/main/java/org/comixedproject/service/comic/ComicService.java:156]: record is never used now.

File: comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateHandler.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                      public void postStateChange(
+                          final State<ComicState, ComicEvent> state,
+                          final Message<ComicEvent> message,
+                          final Transition<ComicState, ComicEvent> transition,
+                          final StateMachine<ComicState, ComicEvent> stateMachine,
+                          final StateMachine<ComicState, ComicEvent> rootStateMachine) {
+                        listeners.forEach(listener -> listener.onComicStateChange(state, message));
+                      }
+                    });
+              }
```
Comment (BRUCELLA2) [inline comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateHandler.java]: Couldn't we replace this code with something like this: 
`
   this.stateMachine.getStateMachineAccessor().doWithAllRegions(

                function -> function.addStateMachineInterceptor(

                    new StateMachineInterceptorAdapter<>() {

                      @Override
                      public void postStateChange(

                          final State<ComicState, ComicEvent> state,
                          final Message<ComicEvent> message,
                          final Transition<ComicState, ComicEvent> transition,
                          final StateMachine<ComicState, ComicEvent> stateMachine,
                          final StateMachine<ComicState, ComicEvent> rootStateMachine) {
                        listeners.forEach(listener -> listener.onComicStateChange(state, message));

                      }
                    }));`

File: comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateHandler.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            });
+  }
+
+  /**
+   * Adds a listener.
+   *
+   * @param listener the listener
+   */
+  public void addListener(final ComicStateChangeListener listener) {
+    log.debug("Adding listener:", listener);
```
Comment (BRUCELLA2) [inline comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateHandler.java]: missing {} ?
log.debug("Adding listener: {}", listener);

File: pom.xml [hunk 4]
```diff
@@ -237,29 +237,7 @@
...
-                    </limit>
-                    <limit>
-                      <counter>CLASS</counter>
-                      <value>MISSEDCOUNT</value>
-                      <maximum>0</maximum>
-                    </limit>
-                  </limits>
-                </rule>
-                -->
+                </rule>                <!-- uncomment when testing has been brought up to 90%                <rule>                  <element>CLASS</element>                  <limits>                    <limit>                      <counter>LINE</counter>                      <value>COVEREDRATIO</value>                      <minimum>0.90</minimum>                    </limit>                    <limit>                      <counter>BRANCH</counter>                      <value>COVEREDRATIO</value>                      <minimum>0.90</minimum>                    </limit>                    <limit>                      <counter>CLASS</counter>                      <value>MISSEDCOUNT</value>                      <maximum>0</maximum>                    </limit>                  </limits>                </rule>                -->
```
Comment (BRUCELLA2) [inline pom.xml]: the indentation has been broken.

File: comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateMachineConfig.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+import org.comixedproject.model.comic.ComicState;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.statemachine.config.EnableStateMachine;
+import org.springframework.statemachine.config.EnumStateMachineConfigurerAdapter;
+import org.springframework.statemachine.config.builders.StateMachineStateConfigurer;
+import org.springframework.statemachine.config.builders.StateMachineTransitionConfigurer;
+
+/**
+ * <code>ComicStateMachineConfig</code> provides a state machine configuration to manage the state
+ * for instances of {@link Comic}.
```
Comment (BRUCELLA2) [inline comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateMachineConfig.java:32]: Exit code: 1 - /home/runner/work/comixed/comixed/comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateMachineConfig.java:31: error: reference not found
Error:   * for instances of {@link Comic}.

-> Import missing : import org.comixedproject.model.comic.Comic;

File: comixed-services/src/main/java/org/comixedproject/service/comic/ComicService.java (hunk 1 above)
Comment (mcpierce) [reply comixed-services/src/main/java/org/comixedproject/service/comic/ComicService.java:156]: Fixed.

File: comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateHandler.java (hunk 2 above)
Comment (mcpierce) [reply comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateHandler.java]: Definitely can. I'm still adapting to the newer (to me) Java syntax and appreciate any point where you can point these out to me. 👍

File: comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateHandler.java (hunk 3 above)
Comment (mcpierce) [reply comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateHandler.java]: Fixed.

File: pom.xml (hunk 4 above)
Comment (mcpierce) [reply pom.xml]: Fixed.

File: comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateMachineConfig.java (hunk 5 above)
Comment (mcpierce) [reply comixed-state/src/main/java/org/comixedproject/state/comic/ComicStateMachineConfig.java:32]: Fixed.

**Comment by BRUCELLA2** [comment]
there are errors in the build :
Error:  Tests run: 6, Failures: 0, Errors: 6, Skipped: 0, Time elapsed: 9.104 s <<< FAILURE! - in org.comixedproject.repositories.library.LastReadRepositoryTest
Error:  testGetEntriesForUserNoEntries  Time elapsed: 0.458 s  <<< ERROR!

When i try to run this branch on my local, I have errors:
2021-04-30 11:06:46.020  WARN 13780 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'liquibase' defined in class path resource [org/springframework/boot/autoconfigure/liquibase/LiquibaseAutoConfiguration$LiquibaseConfiguration.class]: Invocation of init method failed; nested exception is liquibase.exception.ValidationFailedException: Validation Failed:
     1 change sets check sum
          /db/migrations/0.9.0/002_593_migrate_to_camel_case.xml::002_593_migrate_to_camel_case.xml::mcpierce was: 8:5eaf7cdd8a60b135ae295e0090dc133c but is now: 8:24058a0e83a9cd6130d0a273f0aafd2f

**Comment by mcpierce** [comment]
@BRUCELLA2 That's because the migration has changed. Since this change is still a work in progress, to re-run it you'll want to backup you database, run this to test, and then restore your database backup afterward.

**Pull request closed without merging**